applications/crossbar/src/modules/cb_functions.erl

```diff
@@ -0,0 +1,386 @@
+    _ = crossbar_bindings:bind(<<"*.allowed_methods.functions">>, ?MODULE, 'allowed_methods'),
+    _ = crossbar_bindings:bind(<<"*.resource_exists.functions">>, ?MODULE, 'resource_exists'),
+    _ = crossbar_bindings:bind(<<"*.content_types_provided.functions">>, ?MODULE, 'content_types_provided'),
+    _ = crossbar_bindings:bind(<<"*.content_types_accepted.functions">>, ?MODULE, 'content_types_accepted'),
+    _ = crossbar_bindings:bind(<<"*.validate.functions">>, ?MODULE, 'validate'),
+    _ = crossbar_bindings:bind(<<"*.billing">>, ?MODULE, 'billing'),
+    _ = crossbar_bindings:bind(<<"*.execute.get.functions">>, ?MODULE, 'get'),
+    _ = crossbar_bindings:bind(<<"*.execute.put.functions">>, ?MODULE, 'put'),
+    _ = crossbar_bindings:bind(<<"*.execute.post.functions">>, ?MODULE, 'post'),
+    _ = crossbar_bindings:bind(<<"*.execute.patch.functions">>, ?MODULE, 'patch'),
+    _ = crossbar_bindings:bind(<<"*.execute.delete.functions">>, ?MODULE, 'delete'),
+    _ = crossbar_bindings:bind(<<"*.finish_request">>, ?MODULE, 'finish_request'),
+    ok.
+
+%%------------------------------------------------------------------------------
+%% @doc Authenticates the incoming request, returning true and context
+%% if the requestor is known, or false if not.
+%% @end
+%%------------------------------------------------------------------------------
+-spec authenticate(cb_context:context()) ->
+                          {'true', cb_context:context()} |
+                          'false'.
+authenticate(Context) ->
+    authenticate(Context, cb_context:req_verb(Context), cb_context:req_nouns(Context)).
+
+-spec authenticate(cb_context:context(), http_method(), req_nouns()) ->
+                          {'true', cb_context:context()} |
+                          'false'.
+authenticate(Context, ?HTTP_GET, [{<<"functions">>, []}]) ->
+    {'true', Context};
+authenticate(Context, ?HTTP_GET, [{<<"functions">>, [?PATH_TOKEN_SAMPLES | _]}]) ->
+    {'true', Context};
+authenticate(Context, ?HTTP_PUT, [{<<"functions">>, _}]) ->
+    {'true', Context};
+authenticate(Context, ?HTTP_PATCH, [{<<"functions">>, _}]) ->
+    {'true', Context};
+authenticate(Context, ?HTTP_DELETE, [{<<"functions">>, _}]) ->
+    {'true', Context};
+authenticate(_Context, _Verb, _Nouns) -> 'false'.
+
+%%------------------------------------------------------------------------------
+%% @doc Authorizes the incoming request, returning true if the requestor is
+%% allowed to access the resource, or false if not.
+%% @end
+%%------------------------------------------------------------------------------
+-spec authorize(cb_context:context()) -> boolean().
+authorize(Context) ->
+    authorize(Context
+             ,cb_context:req_nouns(Context)
+             ,cb_context:req_verb(Context)
+             ).
+
+-spec authorize(cb_context:context(), req_nouns(), http_method()) -> boolean().
+authorize(Context, [{<<"functions">>, _}], ?HTTP_POST) ->
```

Comment:
What is `POST /v2/functions/` used for?

Reply:
Also, `allowed_methods` says POST only allowed with a function ID present on the URI.

In general, avoid `_` matches unless it really does collapse a lot of identical clauses. Here I'd still like to explicitly match `{<<"functions">>, []}` and other variations.

Reply:
We will only support PATCH and removed POST